core/base/v7/inc/ROOT/RError.hxx

```diff
@@ -0,0 +1,202 @@
+*/
+// clang-format on
+class RStatusTypeSyscall : public RStatusType<int> {
+public:
+   RStatusTypeSyscall() = default;
+   explicit RStatusTypeSyscall(int value) : RStatusType<int>(value) {}
+   bool IsError() const { return fValue < 0; }
+   void SetError() { fValue = -1; }
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RStatusBase
+\ingroup Base
+\brief Type-independent logic of the RStatus wrapper
+*/
+// clang-format on
+class RStatusBase {
+protected:
+   static thread_local bool fgThrowInstantExceptions; /* true by default */
+public:
+   static void SetThrowInstantExceptions(bool value) { fgThrowInstantExceptions = value; }
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RStatus
+\ingroup Base
+\brief Wraps a return value such that unchecked error states trigger an exception
+
+The Status wrapper itself is movable but not copyable to prevent multiple exceptions from being thrown.
+The wrapper is also only stack allocatable because throwing an exception depends on the object going out of scope.
+
+If the class should be used like a C return type, it has to be cleared by a call to IsError() or IsValid() or
+ClearError().  Otherwise, an error state will throw an exception nevertheless when the object goes out of scope.
+*/
+// clang-format on
+template <typename T> // Has to be RStatusType
+class RStatus : public RStatusBase {
+   T fStatus;
+   bool fIsChecked = false;
+
+public:
+   // Named constructor for error cases
+   static RStatus Fail(const std::string &why)
+   {
+      RStatus status;
+      status.MayThrow(why);
+      return status;
+   }
+
+   RStatus() = default;
+   RStatus(const RStatus &other) = delete;
+   RStatus(RStatus &&other) = default;
+   RStatus &operator =(const RStatus &other) = delete;
+   RStatus &operator =(RStatus &&other) = default;
+
+   // Construct from return value
+   explicit RStatus(const typename T::ValueType_t &value) : fStatus(value) {}
+   // Assign a return value
+   RStatus &operator =(const typename T::ValueType_t &value) { this->fStatus = value; }
+
+   ~RStatus() noexcept(false)
+   {
+      if (!fIsChecked && fStatus.IsError()) {
+         // Prevent from throwing if the object is deconstructed in the course of stack unwinding for another exception
+#if __cplusplus >= 201703L
+         if (std::uncaught_exceptions() == 0)
+#else
+         if (!std::uncaught_exception())
+#endif
+         {
+            throw RException("unchecked error");
+         }
+      }
+   }
+
+   void MayThrow(const std::string &why)
+   {
+      fStatus.SetError();
+      // Fast path reserved for return code checking
+      if (R__unlikely(fgThrowInstantExceptions))
+         throw RException(why);
+   }
+
+   bool IsError()
+   {
+      fIsChecked = true;
+      return fStatus.IsError();
+   }
+   bool IsValid() { return !IsError(); }
+   void ClearError() { fIsChecked = true; }
+
+   operator typename T::ValueType_t() const { return fStatus.Get(); }
```

Comment:
Should using this operator count as 'checking' the error status?

Reply:
The operator doesn't check but is currently only supposed to allow for using the RStatus object in expressions as if it was the wrapped type.  I thought it would make sense to require users of RStatus to explicitly spell out "checking" by calling either of `IsError()`, `IsValid()`, `ClearError()`.

Reply:
I don't understand your problem.  It seems with the current implementation the following will *unexpectedly* throw:
```
if (TestSucess(...)) {
   // go on
} else {
   // Try alternative path
}
```
or am I missing something?

Reply:
When `RStatus` captures a boolean, the correct code is admittedly not super intuitive:

    auto rv = TestSucess(...);
    if (rv) {
       rv.ClearError();
       // go on
    } else {
       // alternative
    }

What I wanted to avoid, however, is that just reading the wrapped value counts as checking, such as in

    int filedes = Open("/no/such/file"); // The returned RStatus decays into an int, has been automatically "checked" but filedes contains -1

Reply:
In your last example, the compiler will (sometimes :( ) complains that the value is assigned but not used ...

Reply:
humm ... it sounds like the pattern this code recommends is something like:
```
auto result = Open("/no/such/file");
if (result.IsValid()) {
   int filedesc = result.Get();
   // go on
} else {
   // print error message.
}
```
if that is the case, do we really need the conversion operator to int?  Can we instead have just a operator bool().  Then the 'only' broken case would be:
```
bool result = Open("/no/such/file");
// don't use result at all ...
```

Reply:
@jblomer Sorry.  I am not sure (i.e. I don't remember) how we settle that discussion.  Could you add a comment?

Reply:
It's now modified such that the `RResult` class's `bool` operator is used for checking for error/success. In order to get the encapsulated value, an explicit call the `RResult<T>::Get()` is required. There is now also support for `RResult<void>`, i.e. for functions that don't return anything but still can go wrong.